corehq/form_processor/submission_post.py

```diff
@@ -221,25 +234,34 @@ def run(self):
         if failure_response:
             return FormProcessingResult(failure_response, None, [], [], 'known_failures')
 
+        try:
+            # decode xml to ensure it can be decoded to string when and if needed in pre processing steps
+            instance_xml = self.instance.decode() if isinstance(self.instance, bytes) else self.instance
+        except UnicodeDecodeError as e:
+            return get_submission_error(self.domain, self.instance, e, self.auth_context.to_json())
+
+        xform_context = SubmissionFormContext(instance_xml=instance_xml)
+        form_processing_result = self._pre_process_form(xform_context)
+        if form_processing_result:
+            return form_processing_result
+        # encode xml back
+        self.instance = xform_context.instance_xml.encode()
```

Comment:
this seems weird, why is it necessary?

Reply:
I have had back and forth about this in my head too.

The moment i converted xml to string for regex matches and then retained the string version, other places started failing, https://travis-ci.org/github/dimagi/commcare-hq/jobs/711994384 so i put it back into its original type.
Open to suggestions if there are better ways to do this.

Reply:
How far did you go trying to resolve the errors? It seems like it would be better not to have to do this.

Reply:
I actually didn't try to fix it. I was worried about changing the current expectations of the processes after pre-processing was done. There is a lot happening in this method and around form processing (like saving form xml as attachment) which I was concerned about breaking. Do you see this(encode-> decode) causing any issue or for any reason that it should not be done? I did assume that encode will encode the form back reliably.

Reply:
hey @snopoke , to get more confidence on this,
1. I reached out to Daniel to get his thoughts on this as well. he said
> I'm not sure if it will work to decode the form as UTF-8, then do the find/replace, and finally re-encode as UTF-8? That seems like the simplest approach. However, if some forms are not encoded as UTF-8 then you will run into problems and more complex processing will be necessary, which may involve parsing the XML just enough to extract the encoding.

2. and then i reached out to shubham to confirm
> Mobile only use utf-8 to serialize xform

So we should be good here or you suggest we should try and fix the things that fail?

I need to do something similar for https://dimagi-dev.atlassian.net/browse/ICDS-1682 where I need the decoded version of the form xml, also which sounds like a good candidate for a pre processing step but for all environments and not just ICDS.

Reply:
It seems like there are 2 issues here:

1. How reliably we can decode the raw XML without parsing it
2. If we should re-encode it afterwards

I'm not going to get into 1 since it seems like Daniel has already spoken to that.

Re 2. you are already checking if the XML is bytes and only decoding if it is. That tells me that we already support processing decoded strings so I don't see the point of trying to re-encode it again afterwards. If there is some place that's expecting the raw XML to be `bytes` then it's probably a bug.

Reply:
I think the next (but possibly not only) place where instance_xml is consumed is [`xml2json`](https://github.com/dimagi/xml2json/blob/1b5b4d80b02682007e9aad7893cf2296fe3b58a7/xml2json/lib.py#L15), which calls `lxml.etree.fromstring`. The [documentation](https://lxml.de/tutorial.html#the-fromstring-function) is not very clear, but I'm guessing that accepts either `str` or `bytes`. The algorithm used to determine the encoding there (in case of bytes passed in) is probably what should be used to determine which encoding should be used here.

Note that it is not safe to attempt to convert bytes to text here by guessing the encoding and running with it if there are no errors. It's possible to have no errors, but wrong encoding ending up with character soup, which would mean corrupt form data. Here is a trivial example of bytes that decode without errors in two encodings:
```py
>>> x = b'\xc3\xb0'
>>> x.decode('iso-8859-1')
'Ã°'
>>> x.decode('utf-8')
'ð'
```

Reply:
@millerdev it seems like the only 'safe' approach would be to parse this to XML and then convert it back to a string, does that seem right?

Unless we enforce that forms must be `utf8` encoded which is another option but has it's own issues as you point out above.

Reply:
Yes, that does seem like a possible safe option, although given [what we're looking for](https://github.com/dimagi/commcare-hq/blob/4b839e8606a1c0a72cc54c8c7ac5ce8d4acd2149/custom/icds/form_processor/steps.py#L6), seems like it would be much better to look for the XML element (using xpath) and change the value in the parsed content rather than _parse XML, convert to string, do regex find/replace, re-parse XML_.